gadget/update.go

```diff
@@ -326,6 +327,99 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
 	return nil
 }
 
+// DiskVolumeDeviceTraitsForDevice takes a gadget volume and an expected
+// disk device path and builds up the traits for that device. It validates that
+// the specified volume can be laid out exactly to the disk device, returning an
+// error if the two are not compatible.
+func DiskVolumeDeviceTraitsForDevice(laidOutVolume *LaidOutVolume, dev string) (res DiskVolumeDeviceTraits, err error) {
+	vol := laidOutVolume.Volume
+
+	// get the disk layout for this device
+	diskLayout, err := OnDiskVolumeFromDevice(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot read %v partitions for candidate volume %s: %v", dev, vol.Name, err)
+	}
+
+	// ensure that the on disk and the laid out version are actually compatbile
+	opts := &EnsureLayoutCompatibilityOptions{
+		AssumeCreatablePartitionsCreated: true,
+	}
+	if err := EnsureLayoutCompatibility(laidOutVolume, diskLayout, opts); err != nil {
+		return res, fmt.Errorf("volume %s is not compatible with disk %s: %v", vol.Name, dev, err)
+	}
+
+	// also get a Disk{} interface for this device
+	disk, err := disks.DiskFromDeviceName(dev)
+	if err != nil {
+		return res, fmt.Errorf("cannot get disk for device %s: %v", dev, err)
+	}
+
+	mappedStructures := make([]DiskStructureDeviceTraits, 0, len(diskLayout.Structure))
+
+	diskPartitions, err := disk.Partitions()
+	if err != nil {
+		return res, fmt.Errorf("cannot get partitions for disk device %s: %v", dev, err)
+	}
+
+	// make an easy map of device nodes to partitions for lookup
+	diskPartitionsByDeviceNode := map[string]disks.Partition{}
+	for _, p := range diskPartitions {
+		diskPartitionsByDeviceNode[p.KernelDeviceNode] = p
+	}
+
+	for _, structure := range diskLayout.Structure {
+		part, ok := diskPartitionsByDeviceNode[structure.Node]
+		if !ok {
+			// Somehow this structure in the laid out disk does not exist in
+			// the same physical disk we built for it using udev and
+			// disks.DiskFromDeviceName.
+			// This is unexpected as even `type: bare` i.e. raw structures will
+			// show up in the partition table and thus have a device node
+			// associated with it - this is also true of the "BIOS Boot"
+			// structure for example too, so it's unclear how this case could be
+			// triggered in practice.
```

Comment:
If it's bare there's no partition, so no entry in the partition table.

However, we're iterating over structures created by looking at an actual disk, and we ran EnsureCompatibleLayouts() before, which probably complains if there are more partitions than expected.

Reply:
yeah sorry this comment was out-dated from before I had the realization that `type: bare` structures do not even show up in the partition table, and I assumed that a partition without a filesystem was a `type: bare` structure but as you pointed out this is not the case. I will update this comment

Reply:
done